File: src/0007.reverse-integer/main.c
// Copyright (c) 2024 Xu Shaohua <shaohua@biofan.org>. All rights reserved.
// Use of this source is governed by General Public License that can be
// found in the LICENSE file.

#include <assert.h>
#include <limits.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

int reverse(int x) {
  int rev = 0;
  while (x != 0) {
    if (rev > INT_MAX / 10 || rev < INT_MIN / 10) {
      return 0;
    }
    rev = rev * 10 + x % 10;
    x /= 10;
  }

  return rev;
}

void checkSolution(int x, int expected_rev_x) {
  int rev_x = reverse(x);
  printf("x: %d, rev_x: %d, expected: %d\n", x, rev_x, expected_rev_x);
  assert(rev_x == expected_rev_x);
}

int main() {
  checkSolution(123, 321);
  checkSolution(-123, -321);
  checkSolution(120, 21);
  checkSolution(1534236469, 0);
  checkSolution(-2147483648, 0);
  checkSolution(-2147483412, -2143847412);

  return 0;
}

File: src/0007.reverse-integer/src/main.rs
// Copyright (c) 2024 Xu Shaohua <shaohua@biofan.org>. All rights reserved.
// Use of this source is governed by General Public License that can be found
// in the LICENSE file.

#[allow(clippy::manual_range_contains)]
pub fn reverse(x: i32) -> i32 {
    let mut x = x;
    let mut rev = 0;

    // x == 0 时, 表示它的所有整数进位值都被提取完了.
    while x != 0 {
        // 检查 rev 在添加新的个位值后是否会溢出
        if rev > i32::MAX / 10 || rev < i32::MIN / 10 {
            return 0;
        }
        // 从 x 中提取出个位的值, 然后作为新的个位数值, 添加到 rev 上.
        rev = rev * 10 + x % 10;
        x /= 10;
    }
    rev
}

pub type SolutionFn = fn(i32) -> i32;

fn check_solution(func: SolutionFn) {
    const RECORDS: &[(i32, i32)] = &[
        (123, 321),
        (-123, -321),
        (120, 21),
        (1534236469, 0),
        (-2147483648, 0),
        (-2147483412, -2143847412),
    ];
    for record in RECORDS {
        assert_eq!(func(record.0), record.1);
    }
}

fn main() {
    check_solution(reverse);
}

#[cfg(test)]
mod tests {
    use super::{check_solution, reverse};

    #[test]
    fn test_reverse() {
        check_solution(reverse);
    }
}
